txstratum/filters.py

```diff
@@ -0,0 +1,132 @@
+                for line in fp:
+                    line = line.strip()
+                    if not line:
+                        continue
+                    file_filter.log.info('Added to banned tx ids', txid=line)
+                    file_filter.banned_tx_ids.add(bytes.fromhex(line))
+
+        if address_filename:
+            with open(address_filename, 'r') as fp:
+                for line in fp:
+                    line = line.strip()
+                    if not line:
+                        continue
+                    file_filter.log.info('Added to banned addresses', txid=line)
+                    file_filter.banned_addresses.add(line)
+        return file_filter
+
+    async def close(self) -> None:
+        """Nothing to do here."""
+        return
+
+    async def check_tx(self, tx: 'BaseTransaction', data: Any) -> bool:
+        """Return if the tx should be blocked."""
+        if len(self.banned_tx_ids):
+            for txin in tx.inputs:
+                if txin.tx_id in self.banned_tx_ids:
+                    self.log.info('banned-tx', data=data)
+                    return True
+
+        if len(self.banned_addresses):
+            for txout in tx.outputs:
+                p2pkh = P2PKH.parse_script(txout.script)
+                if p2pkh is not None:
+                    self.log.info('p2pkh.address', address=p2pkh.address)
+                    if p2pkh.address in self.banned_addresses:
+                        self.log.info('banned-address', data=data)
+                        return True
+        return False
+
+
+class TOIFilter(TXFilter):
+    """Filter tx based on the toi service."""
+
+    def __init__(self, client: 'TOIAsyncClient', block: bool = False) -> None:
+        """Init filter."""
+        self.client = client
+        self.block = block
+        self.log = logger.new()
+
+    async def close(self) -> None:
+        """Close toi client."""
+        await self.client.close()
+
+    async def check_tx(self, tx: 'BaseTransaction', data: Any) -> bool:
+        """Return if the tx should be blocked."""
+        txs: Set[str] = set()
+        addrs: Set[str] = set()
+        for txin in tx.inputs:
+            txs.add(txin.tx_id.hex())
+
+        for txout in tx.outputs:
+            p2pkh = P2PKH.parse_script(txout.script)
+            if p2pkh is not None:
+                self.log.debug('p2pkh.address', address=p2pkh.address)
+                addrs.add(p2pkh.address)
+
+        try:
+            resp = await self.client.check_blacklist(tx_ids=list(txs), addresses=list(addrs))
+            if resp.blacklisted:
+                self.log.info('banned', data=data, issues=resp.issues)
+            return resp.blacklisted
+        except TOIError as ex:
+            # TOI Service error
+            self.log.error("toi_service_error", data=data, body=str(ex))
+            if self.block:
+                # We are rejecting a tx based on a toi faillure
+                # TODO: alert
```

Comment:
Including a "[ALERT]" string on the message is the old method of alerting, the new method will be implemented on #39